cpp/src/arrow/compute/registry.cc

```diff
@@ -34,78 +34,144 @@ namespace compute {
+                                        bool allow_overwrite = false) {
+    return (parent_ == nullptr
+                ? Status::OK()
+                : parent_->CanAddFunctionOptionsType(options_type, allow_overwrite)) &
+           DoAddFunctionOptionsType(options_type, allow_overwrite, /*add=*/true);
   }
 
-  Result<std::shared_ptr<Function>> GetFunction(const std::string& name) const {
+  virtual Result<std::shared_ptr<Function>> GetFunction(const std::string& name) const {
     auto it = name_to_function_.find(name);
     if (it == name_to_function_.end()) {
+      if (parent_ != nullptr) {
+        return parent_->GetFunction(name);
+      }
       return Status::KeyError("No function registered with name: ", name);
     }
     return it->second;
   }
 
-  std::vector<std::string> GetFunctionNames() const {
+  virtual std::vector<std::string> GetFunctionNames() const {
     std::vector<std::string> results;
+    if (parent_ != nullptr) {
+      results = parent_->GetFunctionNames();
+    }
     for (auto it : name_to_function_) {
       results.push_back(it.first);
     }
     std::sort(results.begin(), results.end());
     return results;
   }
 
-  Result<const FunctionOptionsType*> GetFunctionOptionsType(
+  virtual Result<const FunctionOptionsType*> GetFunctionOptionsType(
       const std::string& name) const {
     auto it = name_to_options_type_.find(name);
     if (it == name_to_options_type_.end()) {
+      if (parent_ != nullptr) {
+        return parent_->GetFunctionOptionsType(name);
+      }
       return Status::KeyError("No function options type registered with name: ", name);
     }
     return it->second;
   }
 
-  int num_functions() const { return static_cast<int>(name_to_function_.size()); }
+  virtual int num_functions() const {
+    return (parent_ == nullptr ? 0 : parent_->num_functions()) +
+           static_cast<int>(name_to_function_.size());
+  }
 
  private:
+  Status DoAddFunction(std::shared_ptr<Function> function, bool allow_overwrite,
+                       bool add) {
+#ifndef NDEBUG
+    // This validates docstrings extensively, so don't waste time on it
+    // in release builds.
+    RETURN_NOT_OK(function->Validate());
+#endif
+
+    std::lock_guard<std::mutex> mutation_guard(lock_);
+
+    const std::string& name = function->name();
+    auto it = name_to_function_.find(name);
+    if (it != name_to_function_.end() && !allow_overwrite) {
+      return Status::KeyError("Already have a function registered with name: ", name);
+    }
+    if (add) {
+      name_to_function_[name] = std::move(function);
+    }
+    return Status::OK();
+  }
+
+  Status DoAddAlias(const std::string& target_name, const std::string& source_name,
+                    bool add) {
+    std::lock_guard<std::mutex> mutation_guard(lock_);
+
+    // following invocation must not acquire the mutex
+    ARROW_ASSIGN_OR_RAISE(auto func, GetFunction(source_name));
+    if (add) {
```

Comment:
Should we be doing a check here to see if a function with `target_name` is already registered (would probably also necessitate `allow_overwrite`)?